packages/gatsby/src/datastore/lmdb/query/filter-using-index.ts

```diff
@@ -0,0 +1,640 @@
+  for (let { start, end } of ranges) {
+    start = [keyPrefix, ...start]
+    end = [keyPrefix, ...end]
+    const range = !reverse
+      ? { start, end, limit, offset, snapshot: false }
+      : { start: end, end: start, limit, offset, reverse, snapshot: false }
+    // console.log(`range`, range)
+    // @ts-ignore
+    yield* indexes.getRange(range)
+  }
+}
+
+function* performFullScan(context: IFilterArgs): Generator<IIndexEntry> {
+  // *Caveat*: our old query implementation was putting undefined and null values at the end
+  //   of the list when ordered ascending. But lmdb-store keeps them at the top.
+  //   So in LMDB case, need to concat two ranges to conform to our old format:
+  //     concat(undefinedToEnd, topToUndefined)
+  const {
+    databases: { indexes },
+    reverse,
+    indexMetadata: { keyPrefix },
+  } = context
+  // console.log(`full scan`)
+
+  let start: RangeBoundary = [keyPrefix, getValueEdgeAfter(undefinedSymbol)]
+  let end: RangeBoundary = [getValueEdgeAfter(keyPrefix)]
+  let range = !reverse
+    ? { start, end, snapshot: false }
+    : { start: end, end: start, reverse, snapshot: false }
+
+  // console.log(`full-scan-range1`, range)
+
+  const undefinedToEnd = range
+
+  // Concat null/undefined values
+  end = start
+  start = [keyPrefix, null]
+  range = !reverse
+    ? { start, end, snapshot: false }
+    : { start: end, end: start, reverse, snapshot: false }
+
+  const topToUndefined = range
+
+  // console.log(`full-scan-range2`, range, Array.from(topToUndefined))
+  if (!reverse) {
+    // @ts-ignore
+    yield* indexes.getRange(undefinedToEnd)
+    // @ts-ignore
+    yield* indexes.getRange(topToUndefined)
+  } else {
+    // @ts-ignore
+    yield* indexes.getRange(topToUndefined)
+    // @ts-ignore
+    yield* indexes.getRange(undefinedToEnd)
+  }
+}
+
+/**
+ * Takes results after the index scan and tries to filter them additionally with unused parts of the query.
+ *
+ * This is O(N) but the advantage is that it uses data available in the index.
+ * So it effectively bypasses the `getNode()` call for such filters (with all associated deserialization complexity).
+ *
+ * Example:
+ *   Imagine the index is: { foo: 1, bar: 1 }
+ *
+ * Now we run the query:
+ *   sort: [`foo`]
+ *   filter: { bar: { eq: `test` }}
+ *
+ * Initial filtering pass will have to perform a full index scan (because `bar` is the last field in the index).
+ *
+ * But we still have values of `bar` stored in the index itself,
+ * so can filter by this value without loading the full node contents.
+ */
+function narrowResultsIfPossible(
+  context: IFilterContext,
+  entries: GatsbyIterable<IIndexEntry>
+): GatsbyIterable<IIndexEntry> {
+  const { indexMetadata, dbQueries, usedQueries } = context
+
+  const indexFields = new Map<string, number>()
+  indexMetadata.keyFields.forEach(([fieldName], positionInKey) => {
+    indexFields.set(fieldName, positionInKey + 1) // +1 to offset index id at the beginning of the index
```

Comment:
Why do we offset this? Adding a comment would be helpful

Reply:
I was hoping this comment will help 😅

Basically, every index key is `[indexId, field1, field2, ...]`. So `field1` is at position `0` in `keyFields` array but will be at position `1` in the actual index key (so we add offset because of the `index id` in the first column)

Let me know which comment would be more useful/clear for you.

Reply:
Yeah, I think something more verbose like:

```js
// Every index key is [indexId, field1, field2, ...] and indexMetadata.keyFields will be the field1, field2, etc.
// As indexId is in the first column the fields need to be offset by +1
```